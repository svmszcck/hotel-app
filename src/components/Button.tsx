import React from "react";
import { Button, GestureResponderEvent } from "react-native";

type ButtonVariant = "primary" | "secondary" | "danger";

type ButtonProps = {
  title: string;
  onPress: (event: GestureResponderEvent) => void;
  variant?: ButtonVariant;
};

const CustomButton = ({ onPress, title, variant = "primary" }: ButtonProps) => {
  return (
    <Button
      onPress={onPress}
      title={title}
      color="#841584"
      accessibilityLabel="Learn more about this purple button"
    />
  );
};

export default CustomButton;
