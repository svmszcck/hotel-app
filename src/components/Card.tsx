import React, { FC, ReactNode } from "react";
import { Text, View, StyleSheet, ViewStyle } from "react-native";

import Colors from "constants/colors";
import Spacing from "constants/spacing";

type CardProps = {
  text: string;
  style?: ViewStyle;
};

const Card: FC<CardProps> = ({ text, style }) => {
  return (
    <View style={StyleSheet.flatten([styles.container, style])}>
      <Text>{text}</Text>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    alignSelf: "flex-start",
    padding: Spacing.MEDIUM,
    backgroundColor: Colors.GREEN,
  },
});

export default Card;
