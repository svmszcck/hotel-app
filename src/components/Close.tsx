import React, { FC, useState, useEffect } from "react";
import {
  Text,
  StyleSheet,
  GestureResponderEvent,
  TouchableOpacity,
} from "react-native";

import globalStyles from "global-styles";

type CloseProps = {
  onPress: (event: GestureResponderEvent) => void;
};

const Close: FC<CloseProps> = ({ onPress }) => {
  const [close, showClose] = useState<boolean>(false);

  useEffect(() => {
    setTimeout(() => showClose(true), 500);
  }, []);

  return close ? (
    <TouchableOpacity
      style={StyleSheet.flatten([globalStyles.close, { top: 15, left: 15 }])}
      onPress={onPress}
      activeOpacity={0.8}
    >
      <Text>X</Text>
    </TouchableOpacity>
  ) : null;
};

export default Close;
