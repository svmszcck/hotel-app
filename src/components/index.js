export { default as Slider } from "./Slider";
export { default as Button } from "./Button";
export { default as Card } from "./Card";
export { default as Map } from "./Map";
export { default as HotelPanel } from "./HotelPanel";
export { default as Separator } from "./Separator";
export { default as ScreenContainer } from "./ScreenContainer";
export { default as Close } from "./Close";
export { default as SafeAreaView } from "./SafeAreaView";
export { default as Score } from "./Score";
export { default as Maintenance } from "./Maintenance";
export { default as Carousel } from "./Carousel";
