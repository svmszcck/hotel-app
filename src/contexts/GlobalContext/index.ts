import { createContext } from "use-context-selector";

const AppContext = createContext(null as any);

export default AppContext;
