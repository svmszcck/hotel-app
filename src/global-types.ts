// State Types

export type GlobalState = {
  loading: boolean;
  properties: any[];
  activeProperty: any;
};

export type Action = {
  type: string;
  payload: any;
};
