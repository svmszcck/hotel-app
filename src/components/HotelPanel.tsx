import React, { FC } from "react";
import { Image, View, Text, StyleSheet, Pressable } from "react-native";
import { useNavigation } from "@react-navigation/native";
import Ionicon from "@expo/vector-icons/Ionicons";
import { SharedElement } from "react-navigation-shared-element";

import Colors from "constants/colors";
import Spacing from "constants/spacing";
import Routes from "constants/routes";
import { Separator, Score } from "components";
import usePropertySelect from "hooks/usePropertySelect";

import placeholder from "assets/images/placeholder.png";

const HotelPanel: FC = () => {
  const navigation = useNavigation();
  const activeProperty = usePropertySelect();
  const coverImage = activeProperty?.images[0]?.url;

  return (
    <Pressable
      style={styles.container}
      testID="hotel-panel-container"
      onPress={() => navigation.navigate(Routes.DETAILS as never)}
    >
      <SharedElement id="hotel-photo-0">
        <View style={styles.imageWrapper}>
          <Image
            source={coverImage ? { uri: coverImage } : placeholder}
            style={styles.image}
            resizeMode="cover"
          />
        </View>
      </SharedElement>
      <Score style={styles.score} />
      <View style={styles.info}>
        <Text style={styles.title} numberOfLines={1}>
          {activeProperty?.name}
        </Text>
        <View style={styles.location}>
          <Ionicon
            name="location-outline"
            size={20}
            color={Colors.ORANGE}
            style={styles.locationIcon}
          />
          <Text numberOfLines={2}>
            {activeProperty?.distance.toFixed(1)} km from city center
          </Text>
        </View>
        <Separator style={styles.separator} />
        {activeProperty?.lowest_price_per_night ? (
          <Text>
            From{" "}
            <Text style={styles.price}>
              {activeProperty?.lowest_price_per_night}â‚¬/Night
            </Text>
          </Text>
        ) : (
          <Text>No Price Info</Text>
        )}
      </View>
    </Pressable>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    flex: 1,
  },
  imageWrapper: {
    backgroundColor: Colors.BLACK,
  },
  image: {
    width: 100,
    height: "100%",
    opacity: 0.6,
  },
  info: {
    flex: 1,
    padding: Spacing.MEDIUM,
  },
  location: {
    flexDirection: "row",
    alignItems: "center",
    marginTop: Spacing.MEDIUM,
  },
  locationIcon: {
    marginRight: Spacing.SMALL,
  },
  title: {
    fontSize: 18,
    fontWeight: "500",
  },
  price: {
    color: Colors.ORANGE,
    fontWeight: "600",
  },
  separator: {
    marginBottom: Spacing.MEDIUM,
  },
  score: {
    left: 30,
    right: undefined,
  },
});

export default HotelPanel;
