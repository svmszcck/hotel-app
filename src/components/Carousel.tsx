import React, { useMemo, FC } from "react";
import { Image, View, StyleSheet } from "react-native";
import { SwiperFlatList } from "react-native-swiper-flatlist";
import { SharedElement } from "react-navigation-shared-element";

import { screenWidth } from "utils/ui";
import { trimImages } from "utils/general";
import { PropertyImg } from "global-types";

import placeholder from "assets/images/placeholder.png";

const width = screenWidth();

type CarouselProps = {
  images: PropertyImg[];
};

const Carousel: FC<CarouselProps> = ({ images }) => {
  const trimmedImages = useMemo(() => trimImages(images), [images]);

  return (
    <View style={styles.container}>
      <SwiperFlatList
        showPagination
        data={trimmedImages}
        keyExtractor={(item) => item.url}
        renderItem={({ item }) => {
          return (
            <View style={styles.child} testID="carousel-item">
              <SharedElement id={`hotel-photo-${item.position}`}>
                <Image
                  source={item.url ? { uri: item.url } : placeholder}
                  style={styles.cover}
                  testID="carousel-item-image"
                />
              </SharedElement>
            </View>
          );
        }}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    backgroundColor: "white",
  },
  child: {
    width,
    justifyContent: "center",
  },
  cover: {
    width: "100%",
    height: 250,
    minHeight: 300,
  },
});

export default Carousel;
