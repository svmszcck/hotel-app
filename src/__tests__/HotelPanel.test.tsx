import { render, screen, fireEvent } from "@testing-library/react-native";
import renderer from "react-test-renderer";

import { HotelPanel } from "components";
import GlobalContext from "contexts/GlobalContext";
import { GlobalState } from "global-types";

const mockedNavigate = jest.fn();

jest.mock("@react-navigation/native", () => {
  const actualNav = jest.requireActual("@react-navigation/native");
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockedNavigate,
    }),
  };
});

describe("Hotel Panel Component", () => {
  const initialState: GlobalState = {
    loading: false,
    properties: [
      {
        id: 123,
        external_id: "Twerwe",
        images: [
          "https://i.picsum.photos/id/270/536/354.jpg?hmac=mnhmFHibNPisjwBAA_oFyczU3QOuP0Pd6-oeEm5IQaU",
        ],
        name: "Gdfgdf",
        distance: 1.23234,
        lowest_price_per_night: 75,
      },
    ],
    activeProperty: 123,
  };

  const contextValues = {
    ...initialState,
    setProperties: jest.fn(),
    setActiveProperty: jest.fn(),
  };

  test("renders correctly", () => {
    const tree = (
      <GlobalContext.Provider value={contextValues}>
        <HotelPanel />
      </GlobalContext.Provider>
    );

    render(tree);
    const rendered = renderer.create(tree).toJSON();

    fireEvent.press(screen.getByTestId("hotel-panel-container"));

    expect(mockedNavigate).toHaveBeenCalledTimes(1);
    expect(screen.getByText("Gdfgdf")).toBeTruthy();
    expect(screen.getByText("1.2 km from city center")).toBeTruthy();
    expect(screen.getByText("75â‚¬/Night")).toBeTruthy();
    expect(rendered).toMatchSnapshot();
  });
});
