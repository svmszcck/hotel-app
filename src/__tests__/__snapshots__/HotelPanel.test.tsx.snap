// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hotel Panel Component renders correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flex": 1,
      "flexDirection": "row",
    }
  }
  testID="hotel-panel-container"
>
  <View
    collapsable={false}
  >
    <View
      style={
        Object {
          "backgroundColor": "#000000",
        }
      }
    >
      <Image
        resizeMode="cover"
        source={1}
        style={
          Object {
            "height": "100%",
            "opacity": 0.6,
            "width": 100,
          }
        }
      />
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#F7F1E9",
        "borderRadius": 3,
        "flexDirection": "row",
        "left": 30,
        "paddingHorizontal": 10,
        "paddingVertical": 5,
        "position": "absolute",
        "right": undefined,
        "top": 10,
      }
    }
  >
    <Text>
      4.5
    </Text>
    <Text />
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "padding": 10,
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        Object {
          "fontSize": 18,
          "fontWeight": "500",
        }
      }
    >
      Gdfgdf
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginTop": 10,
        }
      }
    >
      <Text />
      <Text
        numberOfLines={2}
      >
        1.2
         km from city center
      </Text>
    </View>
    <View
      style={
        Object {
          "borderBottomColor": "#000000",
          "borderBottomWidth": 0.5,
          "marginBottom": 10,
          "marginVertical": 10,
          "width": "100%",
        }
      }
    />
    <Text>
      From
       
      <Text
        style={
          Object {
            "color": "#B26422",
            "fontWeight": "600",
          }
        }
      >
        75
        â‚¬/Night
      </Text>
    </Text>
  </View>
</View>
`;
